DiverseSelector/metric.py

```diff
@@ -23,25 +23,292 @@
-    # scipy.spatial.distance.pdist.html?highlight=pdist#scipy.spatial.distance.pdist
-    if metric == "euclidean":
-        arr_dist = cdist(feature, feature, "euclidean")
+    """Compute pairwise distance.
+
+    Parameters
+    ----------
+    feature : ndarray
+        feature matrix
+    metric : str
+        metric to use
+
+    Returns
+    -------
+    arr_dist : ndarray
+        symmetric distance array
+    """
+    return cdist(feature, feature, metric)
+
+
+def distance_to_similarity(distance: np.array):
+    """Compute similarity.
+
+    Parameters
+    ----------
+    distance : ndarray
+        symmetric distance array
+
+    Returns
+    -------
+    similarity : ndarray
+        symmetric similarity array.
+    """
+    similarity = 1 / (1 + distance)
+    return similarity
+
+
+def pairwise_similarity_bit(feature: np.array, metric):
+    """Compute the pairwise similarity coefficients.
+
+    Parameters
+    ----------
+    feature : ndarray
+        feature matrix in bit string
+    metric : str
+        method of calculation
+
+    Returns
+    -------
+    pair_coeff : ndarray
+        similarity coefficients for all molecule pairs in feature matrix
+    """
+    pair_simi = []
+    size = len(feature)
+    for i in range(0, size):
+        for j in range(i + 1, size):
+            pair_simi.append(metric(feature[i], feature[j]))
+    pair_coeff = (squareform(pair_simi) + np.identity(size))
+    return pair_coeff
+
+
+# this section is the similarity metrics for non-bitstring input
+
+# todo: we need to compute the pairwise distance matrix for all the molecules in the matrix
+def tanimoto(a, b):
+    """Compute tanimoto coefficient.
 
-    return arr_dist
+    Parameters
+    ----------
+    a : array_like
+        molecule A's features
+    b : array_like
+        molecules B's features
+
+    Returns
+    -------
+    coeff : int
+        tanimoto coefficient for molecule A and B
+    """
+    coeff = (sum(a * b)) / ((sum(a ** 2)) + (sum(b ** 2)) - (sum(a * b)))
+    return coeff
+
+
+def cosine(a, b):
+    """Compute cosine coefficient.
+
+    Parameters
+    ----------
+    a : array_like
+        molecule A's features
+    b : array_like
+        molecules B's features
+
+    Returns
+    -------
+    coeff : int
+        cosine coefficient for molecule A and B
+    """
+    coeff = (sum(a * b)) / (((sum(a ** 2)) + (sum(b ** 2))) ** 0.5)
+    return coeff
+
+
+def dice(a, b):
+    """Compute dice coefficient.
+
+    Parameters
+    ----------
+    a : array_like
+        molecule A's features
+    b : array_like
+        molecules B's features
+
+    Returns
+    -------
+    coeff : int
+        dice coefficient for molecule A and B
+    """
+    coeff = (2 * (sum(a * b))) / ((sum(a ** 2)) + (sum(b ** 2)))
+    return coeff
+
+
+# this section is bit_string similarity calcualtions
+
+
+def modified_tanimoto():
+    """Compute modified tanimoto coefficient."""
+    pass
+
+
+def bit_tanimoto(a, b):
```

Comment:
What's the difference between `tanimoto` and `bit_tanimoto`?

Reply:
sorry for the confusion, `Tanimoto` is for molecular descriptors while `bit_tanimoto`  is for fingerprints. Should I make it so that there's one function that detects the type of input and then processes it accordingly?